File: 32 Diamond/python/diamond.py
"""
DIAMOND

Prints pretty diamond patterns to the screen.

Ported by Dave LeCompte
"""


def print_with_tab(space_count, msg):
    if space_count > 0:
        spaces = " " * space_count
    else:
        spaces = ""
    print(spaces + msg)


def print_diamond(begin_width, end_width, step, width, count):
    edgeString = "CC"
    fill = "!"

    n = begin_width
    while True:
        line_buffer = " " * ((width - n) // 2)
        for across in range(count):
            for a in range(n):
                if a >= len(edgeString):
                    line_buffer += fill
                else:
                    line_buffer += edgeString[a]
            line_buffer += " " * (
                (width * (across + 1) + (width - n) // 2) - len(line_buffer)
            )
        print(line_buffer)
        if n == end_width:
            return
        n += step


def main():
    print_with_tab(33, "DIAMOND")
    print_with_tab(15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print()
    print()
    print()
    print("FOR A PRETTY DIAMOND PATTERN,")
    print("TYPE IN AN ODD NUMBER BETWEEN 5 AND 21")
    width = int(input())
    print()

    PAGE_WIDTH = 60

    count = int(PAGE_WIDTH / width)

    for down in range(count):
        print_diamond(1, width, 2, width, count)
        print_diamond(width - 2, 1, -2, width, count)

    print()
    print()


if __name__ == "__main__":
    main()

File: 32 Diamond/perl/diamond.pl
#!/usr/bin/perl
use strict;

################
# PORTING NOTES:
# * In basic "Tab" function are not spaces, but absolute col position on screen.
# * It was too dificult to port this one, couldn't figure out the original algorithm.
# * So the algorithm was remake.
#

print ' 'x 33 . "DIAMOND\n";
print ' 'x 15 . "CREATIVE COMPUTING MORRISTOWN, NEW JERSEY\n";
print "\n"; print "\n"; print "\n";
print "FOR A PRETTY DIAMOND PATTERN,\n";
print "TYPE IN AN ODD NUMBER BETWEEN 5 AND 21? "; chomp(my $R = <STDIN>); print "\n";


my $Wid= int(60/$R)+1;
my $Dia="CC". "!" x ($R-2);

for (my $J=1; $J<$Wid; $J++) {
	for (my $K=1; $K<($R+2)*2-4; $K+=2) {
		my $Size= $K;
		if ($K>$R) { $Size=$R+($R-$K); }
		my $Chunk= substr($Dia, 0, $Size);
		for (my $L=1; $L<$Wid; $L++) {
			my $Space= " " x (($R-$Size)/2);
			if ($L>1) { $Space.=$Space; }
			print $Space.$Chunk;
			}
		print "\n";
		}
	}

exit;


